File: Python/data_structures/queue.py
from collections import deque


class Queue:
    def __init__(self, size: int) -> None:
        self.queue = deque(maxlen=size)
        self.max_size = size
        self.size = 0
        self.front = 0
        self.rear = size - 1

    def __iter__(self) -> int:
        for item in self.queue:
            yield item

    def __str__(self) -> str:
        items = []
        for item in self.__iter__():
            items.append(str(item))

        # return a string with each item separated by commas
        if len(items) == 1:
            return "".join(items[0])

        return ", ".join(items[:-1]) + ", " + items[-1] if len(items) > 0 else "Empty"

    def enqueue(self, item: int) -> None:
        if not self.is_full():
            self.queue.append(item)
            self.rear = (self.rear + 1) % self.max_size
            self.size += 1
            print(f"Enqueued item {item} to queue")
            return

        print(f"Queue full, cannot enqueue item {item}")

    def dequeue(self) -> int:
        if not self.is_empty():
            item = self.queue.popleft()
            self.front = (self.front - 1) % self.max_size
            self.size -= 1
            print(f"Dequeued item {item}")
            return item

        print("Queue empty, cannot dequeue item")
        return None

    def is_empty(self) -> bool:
        return self.size == 0

    def is_full(self) -> bool:
        return self.size == self.max_size


if __name__ == "__main__":
    queue = Queue(5)

    queue.enqueue(1)
    print(queue)
    queue.enqueue(2)
    queue.enqueue(3)
    queue.enqueue(4)
    queue.enqueue(5)
    print(queue)
    queue.enqueue(6)
    print(queue)
    queue.dequeue()
    print(queue)
    queue.dequeue()
    queue.dequeue()
    print(queue)
    queue.enqueue(9)
    print(queue)
    queue.dequeue()
    print(queue)
    queue.enqueue(13)
    print(queue)
    queue.dequeue()
    print(queue)
    queue.dequeue()
    queue.dequeue()
    queue.dequeue()
    print(queue)
    queue.enqueue(3)
    queue.enqueue(7)
    queue.enqueue(11)
    print(queue)

File: C++/data_structures/queue.cpp
#include <iostream>
#include <queue>

int main(int argc, char const *argv[])
{
    std::queue<int> queue = std::queue<int>();

    queue.push(122);
    queue.push(145);
    queue.push(109);
    queue.push(156);
    queue.push(134);

    std::cout << "Queue empty: " << queue.empty() << std::endl;
    std::cout << "Queue size: " << queue.size() << std::endl;

    std::cout << "\nQueue dequeued: " << queue.front() << std::endl;
    queue.pop();

    std::cout << "Queue dequeued: " << queue.front() << std::endl;
    queue.pop();

    std::cout << "Queue size: " << queue.size() << std::endl;

    queue.push(156);
    std::cout << "\nQueue enqueued: " << queue.back() << std::endl;

    queue.push(117);
    std::cout << "Queue enqueued: " << queue.back() << std::endl;

    std::cout << "\nQueue: front ->\t" << queue.front();
    queue.pop();

    while (!queue.empty())
    {
        std::cout << "  |  " << queue.front();
        queue.pop();
    }

    std::cout << " <- back" << std::endl;

    return 0;
}
